File: app/src/main/java/xyz/rodit/snapmod/features/FeatureManager.java
package xyz.rodit.snapmod.features;

import java.util.ArrayList;
import java.util.List;

import xyz.rodit.snapmod.features.chatmenu.ChatMenuModifier;
import xyz.rodit.snapmod.features.conversations.MessageInterceptor;
import xyz.rodit.snapmod.features.conversations.PreventBitmojiPresence;
import xyz.rodit.snapmod.features.conversations.PreventReadReceipts;
import xyz.rodit.snapmod.features.conversations.PreventTypingNotifications;
import xyz.rodit.snapmod.features.conversations.SnapInteractionFilter;
import xyz.rodit.snapmod.features.conversations.SnapOverrides;
import xyz.rodit.snapmod.features.friendsfeed.FeedModifier;
import xyz.rodit.snapmod.features.info.AdditionalFriendInfo;
import xyz.rodit.snapmod.features.info.NetworkLogging;
import xyz.rodit.snapmod.features.opera.OperaModelModifier;
import xyz.rodit.snapmod.features.saving.ChatSaving;
import xyz.rodit.snapmod.features.saving.PublicProfileSaving;
import xyz.rodit.snapmod.features.saving.StoriesSaving;
import xyz.rodit.snapmod.features.tweaks.BypassVideoLength;
import xyz.rodit.snapmod.features.tweaks.ConfigurationTweaks;
import xyz.rodit.snapmod.features.tweaks.DisableBitmojis;
import xyz.rodit.snapmod.features.tweaks.HideFriends;
import xyz.rodit.snapmod.features.tweaks.HideStoryReadReceipts;
import xyz.rodit.snapmod.features.tweaks.HideStorySections;
import xyz.rodit.snapmod.features.tweaks.LocationOverride;

public class FeatureManager extends Contextual {

    private final List<Feature> features = new ArrayList<>();

    public FeatureManager(FeatureContext context) {
        super(context);
    }

    public void load() {
        // Chat context menu
        add(ChatMenuModifier::new);

        // Friends feed
        add(FeedModifier::new);

        // Conversations/chats
        add(MessageInterceptor::new);
        add(PreventBitmojiPresence::new);
        add(PreventReadReceipts::new);
        add(PreventTypingNotifications::new);
        add(SnapInteractionFilter::new);
        add(SnapOverrides::new);

        // Opera (story/snap view)
        add(OperaModelModifier::new);

        // Saving
        add(ChatSaving::new);
        add(PublicProfileSaving::new);
        add(StoriesSaving::new);

        // Information
        add(AdditionalFriendInfo::new);
        add(NetworkLogging::new);

        // Tweaks
        add(BypassVideoLength::new);
        add(ConfigurationTweaks::new);
        add(DisableBitmojis::new);
        // add(FriendAddOverride::new);
        add(HideFriends::new);
        add(HideStoryReadReceipts::new);
        add(HideStorySections::new);
        add(LocationOverride::new);
    }

    public void init() {
        features.forEach(Feature::init);
    }

    public void onConfigLoaded(boolean first) {
        features.forEach(f -> f.onConfigLoaded(first));
    }

    public void performHooks() {
        features.forEach(Feature::performHooks);
    }

    public void add(FeatureSupplier supplier) {
        features.add(supplier.supply(context));
    }

    @FunctionalInterface
    public interface FeatureSupplier {

        Feature supply(FeatureContext context);
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/FeatureManager.kt
package xyz.rodit.snapmod.features

import okhttp3.internal.toImmutableList
import xyz.rodit.snapmod.features.chatmenu.ChatMenuModifier
import xyz.rodit.snapmod.features.chatmenu.new.NewChatMenuModifier
import xyz.rodit.snapmod.features.conversations.*
import xyz.rodit.snapmod.features.friendsfeed.PinChats
import xyz.rodit.snapmod.features.info.AdditionalFriendInfo
import xyz.rodit.snapmod.features.info.NetworkLogging
import xyz.rodit.snapmod.features.notifications.FilterTypes
import xyz.rodit.snapmod.features.notifications.ShowMessageContent
import xyz.rodit.snapmod.features.opera.CustomStoryOptions
import xyz.rodit.snapmod.features.opera.OperaModelModifier
import xyz.rodit.snapmod.features.saving.*
import xyz.rodit.snapmod.features.tweaks.*

class FeatureManager(context: FeatureContext) : Contextual(context) {

    private val features: MutableList<Feature> = ArrayList()
    val pubFeatures get() = features.toImmutableList()

    fun load() {
        // Chat context menu
        add(::ChatMenuModifier)
        add(::NewChatMenuModifier)

        // Friends feed
        add(::PinChats)

        // Conversations/chats
        add(::AutoSave)
        add(::MessageInterceptor)
        add(::PreventBitmojiPresence)
        add(::PreventReadReceipts)
        add(::PreventTypingNotifications)
        add(::SnapInteractionFilter)
        add(::SnapOverrides)

        // Notifications
        add(::FilterTypes)
        add(::ShowMessageContent)

        // Opera (story/snap view)
        add(::CustomStoryOptions)
        add(::OperaModelModifier)

        // Saving
        add(::AutoDownloadSnaps)
        add(::AutoDownloadStories)
        add(::ChatSaving)
        add(::PublicProfileSaving)
        add(::StoriesSaving)

        // Information
        add(::AdditionalFriendInfo)
        add(::NetworkLogging)

        // Tweaks
        add(::BypassVideoLength)
        add(::BypassVideoLengthGlobal)
        add(::CameraResolution)
        add(::ConfigurationTweaks)
        add(::ConfigurationTweaks)
        add(::DisableBitmojis)
        // add(::HideFriends)
        add(::HideStoryReadReceipts)
        add(::HideStorySections)
        add(::HideStorySectionsLegacy)
        add(::PinStories)
    }

    fun init() {
        features.forEach { it.init() }
    }

    fun onConfigLoaded(first: Boolean) {
        features.forEach { it.onConfigLoaded(first) }
    }

    fun performHooks() {
        features.forEach { it.performHooks() }
    }

    fun add(supplier: (FeatureContext) -> Feature) {
        val feature = supplier(context)
        if (feature.support.contains(context.appVersion)) {
            features.add(feature)
        }
    }

    inline fun <reified T> get(): T where T : Feature {
        return pubFeatures.filterIsInstance<T>().first()
    }
}